share/wake/lib/core/list.wake

```diff
@@ -802,3 +805,76 @@ export def groupBy (cmp: a => a => Boolean) (list: List a): List (List a) =
             (elem, head, rest), tail
         _ = (elem, Nil), acc
     foldr combo Nil sorted
+
+export def fromOption (x: Option a): List a =
+    getOrElse Nil $ omap (_, Nil) x
+
+# TODO: replace at with atT
+export def atT (i: Integer) (list: List a): List a =
+    at i list | fromOption
+
+export def atE (i: Integer): (a => a) => List a => List a =
+    def edit f l =
+      require Pair first (x, follow) = splitAt i l else l
+      first ++ (f x, follow)
+    edit
```

Comment:
However, I agree that this `atE` method is a hole in the list API. I think it needs a first class name, though. If list already has `at`, ... should this be `editAt` (and we also add a `setAt`)?

Reply:
Hmm. Manipulating lists in this way (one element at a time) would encourage algorithmically bad code. Really, a `map`+`if` loop (ala `editFilter`) that does a bulk change is better.

Reply:
I think I disagree with the whole "edit in bulk" issue here but I've argued myself into ambivalence on the issue so I'm going to make the requested edits leave a "lets keep an eye on it" comment. I think for most cases

Reply:
Actually I think I've argued myself out of ambivalence and to your side on this. I'm not overly swayed by the expectation of n^2 code from people but I am swayed by this

Consider the case where keys are unique vs the case where keys are not unique:
- Most common case: If keys are unique semantically our two ways are the same (mine is a bit more performant due to early stopping but like I said I don't think that matters and its still linear not n^2)
- Next Most common case: If keys are not unique and are not ordered your way is "the one true way"
- Least common case: If keys are not unique and they're *ordered* you _might_ care about getting the first key but many this just seems like such a long shot to me but if this case came up my way would be "the one true way"

So I think in expectation your way is a bit better than my way.